lib/srv/discovery/discovery.go

```diff
@@ -1327,8 +1323,90 @@ func (s *Server) getAllGCPServerFetchers() []server.Fetcher {
 	return allFetchers
 }
 
+// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
+// It will retry using an exponential backoff algorithm.
+func (s *Server) acquireDiscoveryGroup() error {
+	if s.DiscoveryGroup == "" {
+		return nil
+	}
+
+	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
+		First:  0,
+		Driver: retryutils.NewExponentialDriver(defaults.HighResPollingPeriod),
+		Max:    defaults.LowResPollingPeriod,
+		Jitter: retryutils.NewHalfJitter(),
+		Clock:  s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var lease *services.SemaphoreLock
+	for range retry.After() {
+		retry.Inc()
+		s.Log.Debugf("Discovery service is trying to acquire lock for DiscoveryGroup %q", s.DiscoveryGroup)
+		lease, err = s.tryAcquireDiscoveryGroupLease()
+		if err != nil {
+			if !strings.Contains(err.Error(), teleport.MaxLeases) {
+				return trace.Wrap(err)
+			}
+			s.Log.Debugf("Discovery service is waiting on DiscoveryGroup %q lock: %v", s.DiscoveryGroup, err)
+			continue
+		}
+
+		break
+	}
+
+	go func() {
+		for {
+			select {
+			case <-lease.Renewed():
+				continue
+			case <-lease.Done():
+				s.Log.WithError(lease.Wait()).Warnf("DiscoveryGroup %q lock was lost, stopping discovery service", s.DiscoveryGroup)
+				s.Stop()
+				return
+			}
+		}
+	}()
+
+	return nil
+}
```

Comment:
```suggestion

// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
// It will retry using an exponential backoff algorithm.
func (s *Server) runWhileAcquiringDiscoveryGroup(run func(context.Context) error) error {
	if s.DiscoveryGroup == "" {
		s.Log.Warnf("DiscoveryGroup is not set, skipping semaphore lock. It is recommended to set a DiscoveryGroup")
		return trace.Wrap(run(s.ctx))
	}

	s.Log.Debugf("Discovery service is trying to acquire lock for DiscoveryGroup %q", s.DiscoveryGroup)
	lease, err := services.AcquireSemaphoreLock(s.ctx, services.SemaphoreLockConfig{
		Service: s.Config.AccessPoint,
		Expiry:  time.Minute,
		Params: types.AcquireSemaphoreRequest{
			SemaphoreKind: types.SemaphoreKindDiscoveryServiceGroup,
			SemaphoreName: s.DiscoveryGroup,
			MaxLeases:     1,
			Holder:        s.ServerID,
		},
		Clock: s.clock,
	})

	switch {
	case err == nil:
		s.Log.Debugf("Discovery service acquired lock for DiscoveryGroup %q", s.DiscoveryGroup)
		defer lease.Wait()
		defer lease.Stop()
		ctx, cancel := context.WithCancel(s.ctx)
		defer cancel()
		go func() {
			select {
			case <-lease.Done():
				cancel()
			case <-ctx.Done():
				return
			}
		}()
		err := run(ctx)
		switch {
		case errors.Is(err, context.Canceled):
			select {
			case <-s.ctx.Done():
				return trace.Wrap(err)
			default:
				return nil
			}
		}
	case trace.IsAlreadyExists(err) ....:
		s.Log.Debugf("Discovery service is waiting on DiscoveryGroup %q lock", s.DiscoveryGroup)
		return trace.Wrap(err)
	}

	return nil
}

```